libsql_client/dbapi2/types.py

```diff
@@ -0,0 +1,1423 @@
+            tp = self.__class__
+            mod = tp.__module__
+            name = tp.__name__
+            qualname = f"{mod}.{name}"
+            msg = f"cannot create '{qualname}' instances"
+            raise TypeError(msg)
+
+        assert len(args) == 1
+        sql = args[0]
+        if not isinstance(sql, str):
+            raise TypeError("sql must be a str")
+
+        if "\0" in sql:
+            raise ProgrammingError("the query contains a null character")
+        if sys.version_info[:2] >= (3, 11):
+            if len(sql) > con.getlimit(SQLITE_LIMIT_SQL_LENGTH):
+                raise DataError("query string is too large")
+
+        self.sql = sql
+        # see pysqlite_statement_create()
+        # https://github.com/python/cpython/blob/main/Modules/_sqlite/statement.c#L32
+        # however we need to parse the statement with iter_sql_statements()
+        # as we can't use sqlite3_prepare_v2()
+        itr = iter(iter_sql_statements(sql))
+        self.tokens = next(itr)
+        assert self.tokens
+        next_stmt = next(itr, None)
+        if next_stmt and next_stmt[0].upper() != "END":
+            msg = "You can only execute one statement at a time."
+            raise ProgrammingError(msg)
+        self.is_dml = self._check_is_dml(self.tokens)
+
+    @classmethod
+    def _check_is_dml(cls, tokens: Sequence[str]) -> bool:
+        # NOTE: this matches C pysqlite_statement_create()
+        # https://github.com/python/cpython/blob/main/Modules/_sqlite/statement.c#L75-L82
+        # but it's not fully correct as the language allows: WITH ...
+        # See https://www.sqlite.org/lang_insert.html
+        # https://www.sqlite.org/lang_update.html
+        # https://www.sqlite.org/lang_delete.html
+        # Syntax:
+        #    WITH RECURSIVE table_name ( col_name, other_col)
+        #    AS NOT MATERIALIZED ( SELECT ... )
+        return tokens[0].upper() in cls._dml_statements
+
+    @property
+    def is_readonly(self) -> bool:
+        # mimics sqlite3_stmt_readonly()
+        return not self.is_dml
+
+
+def check_thread(fn: Callable[P, T]) -> Callable[P, T]:
+    """Decorator that checks :py:class:`Connection` thread usage.
+
+    Otherwise raises ``ProgrammingError``.
+
+    :meta private:
+    """
+    def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
+        self = args[0]
+        assert len(args) > 0 and isinstance(self, Connection)
+        self._do_check_thread()
+        return fn(*args, **kwargs)
+
+    return wrapper
+
+
+def check_connection(fn: Callable[P, T]) -> Callable[P, T]:
+    """Decorator that checks :py:class:`Connection` being valid.
+
+        Otherwise raises ``ProgrammingError``.
+
+    :meta private:
+    """
+    def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
+        self = args[0]
+        assert len(args) > 0 and isinstance(self, Connection)
+        self._do_check_connection()
+        return fn(*args, **kwargs)
+
+    return wrapper
```

Comment:
What's the point of using a decorator instead of simply calling the `_check` method?

Reply:
I don't understand how

```
def foo(self):
    self._check_connection()
    ...
```

is harder to read and less polluting than

```
@check_connection
def foo(self):
    ...
```

But if it's just a matter of personal preference and there is no functional difference, then why not; you are going to maintain this thing, after all :)

Reply:
the decorator could also change docstrings and the likes, it's a more common pattern